### PR TITLE
Feature request: PercentilePruner

### PR DESCRIPTION
**Pull request by c-bata**
Like MedianPruner, but prune if the best intermediate result is worse than specified percentile in the other trials at the same step.

### CODE REVIEW

File: optuna/storages/base.py [hunk 1]
```diff
@@ -243,6 +243,24 @@ def get_median_intermediate_result_over_trials(self, study_id, step):
...
+
+        all_trials = [
+            t for t in self.get_all_trials(study_id) if t.state == structs.TrialState.COMPLETE
+        ]
+
+        if len(all_trials) == 0:
+            raise ValueError("No trials have been completed.")
+
+        return float(
+            np.nanpercentile(
```
Comment (c-bata) [inline optuna/storages/base.py]: The difference with MedianPruner is essentially this line only.

**Comment by c-bata** [comment]
If MedianPruner inherits PercentilePruner, It might be reduce the redundancy. But I didn't for now. If you want to do it, please tell me.

**Comment by sile** [comment]
Thank you for your nice PR!

> If MedianPruner inherits PercentilePruner, It may be simpler by reducing the redundancy. But I didn't for now. If you want to do it, please tell me.

I want `MedianPruner` to inherit `PercentilePruner`, but I think it's okay either the modification is included in this PR or not.

**Comment by c-bata** [comment]
I see! I've just pushed https://github.com/pfnet/optuna/pull/397/commits/46bf677bb0deb2be12a9d9b9b0d9284149e25674 now.

**Comment by c-bata** [comment]
Rebased this branch on current master.

File: optuna/pruners/percentile.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            >>> def objective(trial):
+            >>>     ...
+            >>>
+            >>> study = create_study(pruner=PercentilePruner())
+            >>> study.optimize(objective)
+
+    Args:
+        percentile:
+            Percentile which must be between 0 and 100 inclusive
+            (ex: When given 25.0, top of 25th percentile trials are kept).
```
Comment (sile) [inline optuna/pruners/percentile.py]: In Optuna documentation, "e.g.," is preferred over "ex:".

**Comment by c-bata** [comment]
Thanks. Fixed at https://github.com/pfnet/optuna/pull/397/commits/ab7aa6f701f4c3452d518c1e3fb0cefd13840cd5

スクリーンショット 2019-05-28 18 33 51

**Comment by g-votte** [comment]
Merged the PR.
Thank you for your contribution! @c-bata

**Pull request merged**